### PR TITLE
Speed up `cupy.vdot`

### LINKED ISSUE #3672: `cupy.vdot` is way slower than naive implementation
I am on CUDA 10.0 + GTX 2080 Ti + master branch, and this is what I see (with `CUPY_ACCELERATORS=cub`):
```python
>>> import cupy as cp
>>> from cupyx.time import repeat
>>>
>>> a = cp.random.random(1000000)
>>> b = cp.random.random(1000000)
>>> def my_vdot(a, b):
...     return cp.sum(a * b)
... 
>>> print(repeat(cp.vdot, (a, b)))
vdot                :    CPU:   17.272 us   +/- 0.938 (min:   16.300 / max:   39.416) us     GPU-0:  640.646 us   +/-18.809 (min:  630.496 / max:  774.144) us
>>> print(repeat(my_vdot, (a, b)))
my_vdot             :    CPU:   26.312 us   +/- 3.240 (min:   25.204 / max:   97.959) us     GPU-0:   74.100 us   +/-12.585 (min:   70.560 / max:  274.432) us
>>> my_vdot(a, b)  == cp.vdot(a, b)
array(True)
```
I could probably look into this myself, just opening a ticket in case I forget and someone can pick up 😅

### PR DESCRIPTION
**Pull request by leofang**
Close #3672. Use CUB device reduction to accelerate it. If not available, fall back to the original kernel.

Before: 
```
vdot                :    CPU:   17.272 us   +/- 0.938 (min:   16.300 / max:   39.416) us     GPU-0:  640.646 us   +/-18.809 (min:  630.496 / max:  774.144) us
```
After:
```
vdot                :    CPU:   57.839 us   +/- 6.615 (min:   55.680 / max:  216.692) us     GPU-0:   80.302 us   +/-18.042 (min:   75.776 / max:  232.096) us
```
For some reason CPU overhead is unexpectedly increased...Not sure what I overlooked 😞

### CODE REVIEW

**Comment by leofang** [comment]
Jenkins, test this please

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -2811,8 +2812,9 @@ cpdef ndarray tensordot_core(
             out = _ndarray_init(ret_shape, dtype)
 
     if m == 1 and n == 1:
-        _tensordot_core_mul_sum(
-            a.ravel(), b.ravel(), _manipulation._reshape(out, ()))
+        # TODO(leofang): switch back to _tensordot_core_mul_sum when
+        # ReductionKernel supports CUB block reduction
+        out = (a.ravel() * b.ravel()).sum(out=_manipulation._reshape(out, ()))
```
Comment (asi1024) [inline cupy/core/core.pyx:2817]: It looks to degrade the performance when CUB is not enabled because your implementation calls kernels twice.

Comment (leofang) [reply cupy/core/core.pyx:2817]: I thought so too, but two kernel launches turned out still faster...Do you know why?

What do you think we add a for loop for routine accelerators here, and then fall back to `_tensordot_core_mul_sum`?

Comment (leofang) [reply cupy/core/core.pyx]: @asi1024 PTAL at 1bd32b5.

**Comment by leofang** [comment]
I am still a bit concerned with the CPU overhead (57 us) which does not show up in my naive implementation (25 us).

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -2812,9 +2812,15 @@ cpdef ndarray tensordot_core(
             out = _ndarray_init(ret_shape, dtype)
 
     if m == 1 and n == 1:
-        # TODO(leofang): switch back to _tensordot_core_mul_sum when
-        # ReductionKernel supports CUB block reduction
-        out = (a.ravel() * b.ravel()).sum(out=_manipulation._reshape(out, ()))
+        if len(_accelerator._routine_accelerators) > 0:
+            # fast path using CUB or cuTENSOR; if not applicable, it'd fall
+            # back to CuPy's reduction kernel
```
Comment (asi1024) [inline cupy/core/core.pyx:2817]: The implementation of this PR fallback to CuPy's default `(a.ravel() * b.ravel()).sum(...)`, not reduction kernel.

Comment (leofang) [reply cupy/core/core.pyx:2817]: Oh, by this I meant `sum` is backed by CuPy's reduction kernel implemented in `cupy/core/_reduction.pyx`, not the `ReductionKernel` object. Would you like me to make it clearer?

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -2812,9 +2812,15 @@ cpdef ndarray tensordot_core(
             out = _ndarray_init(ret_shape, dtype)
 
     if m == 1 and n == 1:
-        # TODO(leofang): switch back to _tensordot_core_mul_sum when
-        # ReductionKernel supports CUB block reduction
-        out = (a.ravel() * b.ravel()).sum(out=_manipulation._reshape(out, ()))
+        if len(_accelerator._routine_accelerators) > 0:
```
[the remark this answers is not in the document]
Comment (leofang) [reply cupy/core/core.pyx:2815]: This logic is handled in `sum()`, isn't it?

[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/core.pyx:2815]: I mean that we may support another accelerator which is not used cupy.sum in the future.

File: cupy/core/core.pyx (hunk 2 above)
Comment (asi1024) [reply cupy/core/core.pyx:2817]: How about `CuPy's default sum kernel`?

File: cupy/core/core.pyx (hunk 3 above)
[the remark this answers is not in the document]
Comment (leofang) [reply cupy/core/core.pyx:2815]: Oh OK, wasn't aware of this possibility. Will fix. 

(What accelerators you think will be added in the near future, btw? Just curious 😇)

File: cupy/core/core.pyx (hunk 2 above)
Comment (leofang) [reply cupy/core/core.pyx:2817]: Sounds good.

File: cupy/core/core.pyx (hunk 3 above)
[the remark this answers is not in the document]
Comment (asi1024) [reply cupy/core/core.pyx:2815]: We have no plan for another accelerator, currently.

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please.

**Comment by asi1024** [comment]
LGTM.

**Pull request merged**